### PR TITLE
add kprobe sample program attached to tcp_connect

### PR DESCRIPTION
**Pull request by dmitris**
PR adds an example kprobe program attached to `tcp_connect`.  It also adds a book page to fill in the current "WIP" https://aya-rs.dev/book/programs/probes/.

Sample output:
```
$ cargo xtask build-ebpf --release && cargo build --release && sudo RUST_LOG=info target/release/kprobetcp

[2022-12-28T15:12:55Z INFO  kprobetcp] Waiting for Ctrl-C...
[2022-12-28T15:13:08Z INFO  kprobetcp] AF_INET6 src addr: 2001:4998:efeb:282:0:0:0:249, dest addr: 2606:2800:220:1:248:1893:25c8:1946
[2022-12-28T15:13:16Z INFO  kprobetcp] AF_INET src address: 10.53.149.148, dest address: 93.184.216.34
[2022-12-28T15:13:19Z INFO  kprobetcp] AF_INET src address: 10.53.149.148, dest address: 10.222.58.32
```

### CODE REVIEW

File: examples/kprobetcp/kprobetcp-ebpf/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                unsafe { u16::from_be(src_addr.in6_u.u6_addr16[7]) },
+                unsafe { u16::from_be(dest_addr.in6_u.u6_addr16[0]) },
+                unsafe { u16::from_be(dest_addr.in6_u.u6_addr16[1]) },
+                unsafe { u16::from_be(dest_addr.in6_u.u6_addr16[2]) },
+                unsafe { u16::from_be(dest_addr.in6_u.u6_addr16[3]) },
+                unsafe { u16::from_be(dest_addr.in6_u.u6_addr16[4]) },
+                unsafe { u16::from_be(dest_addr.in6_u.u6_addr16[5]) },
+                unsafe { u16::from_be(dest_addr.in6_u.u6_addr16[6]) },
+                unsafe { u16::from_be(dest_addr.in6_u.u6_addr16[7]) }
+            );
```
Comment (vadorovsky) [inline examples/kprobetcp/kprobetcp-ebpf/src/main.rs:76]: We have `{:ipv6}` format hint for `[u8; 16]` :) And also, when you use `[u8; 16]` representation of IPv6, you don't have to care about endianess.

```suggestion
            info!(
                &ctx,
                "AF_INET6 src addr: {:ipv6}, dest addr: {:ipv6}",
                src_addr.in6_u.u6_addr8,
                dest_addr.in6_u.u6_addr8,
            );
```

**Review by vadorovsky — CHANGES_REQUESTED** [review]
Thanks for working on this!

I think it would be worth to add this example to this page https://github.com/aya-rs/book/blob/main/docs/book/programs/probes.md

In general, every example should be linked with some book chapter.

File: examples/kprobetcp/kprobetcp-ebpf/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+use crate::binding::{sock, sock_common};
+
+use aya_bpf::{
+    helpers::bpf_probe_read_kernel, macros::kprobe, programs::ProbeContext,
+};
+
+use aya_log_ebpf::info;
+const AF_INET: u16 = 2;
+const AF_INET6: u16 = 10;
```
[the remark this answers is not in the document]
Comment (dmitris) [reply examples/kprobetcp/kprobetcp-ebpf/src/main.rs]: changed

File: examples/kprobetcp/kprobetcp-ebpf/src/main.rs (hunk 1 above)
Comment (dmitris) [reply examples/kprobetcp/kprobetcp-ebpf/src/main.rs]: changed in [bbc14f5](https://github.com/aya-rs/book/pull/81/commits/bbc14f58e8a41908686b42315b441534dacddd8d)

**Comment by dmitris** [comment]
I added a book page describing kprobe programs and linking to the example

File: examples/kprobetcp/kprobetcp/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    let _opt = Opt::parse();
+
+    env_logger::init();
+
+    // This will include your eBPF object file as raw bytes at compile-time and load it at
+    // runtime. This approach is recommended for most real-world use cases. If you would
+    // like to specify the eBPF program at runtime rather than at compile-time, you can
+    // reach for `Bpf::load_file` instead.
+    #[cfg(debug_assertions)]
+    let mut bpfh = Bpf::load(include_bytes_aligned!(
```
Comment (vadorovsky) [inline examples/kprobetcp/kprobetcp/src/main.rs:21]: Where does the `h` in `bpfh` come from? In aya-template, that variable is just called `bpf`.

Comment (dmitris) [reply examples/kprobetcp/kprobetcp/src/main.rs]: changed in 1257df4

File: docs/book/programs/probes.md [hunk 4]
```diff
@@ -1,3 +1,52 @@
 # Probes
 
-This page is a work in progress, please feel free to open a Pull Request!
+!!! example "Source Code"
+
+    Full code for the example in this chapter is availble [here](https://github.com/aya-rs/book/tree/main/examples/kprobetcp).
+
+# What are the probes in eBPF?
+
+The probe BPF programs attach to a kernel (kprobes) or user-side (uprobes) functions and are able to access the function parameters of those functions.  You can find more information about probes in the [kernel documentation](https://docs.kernel.org/trace/kprobes.html), icluding the difference between kprobes and kretprobes.
```
Comment (vadorovsky) [inline docs/book/programs/probes.md:9]:
```suggestion
The probe BPF programs attach to kernel (kprobes) or user-side (uprobes) functions and are able to access the function parameters of those functions.  You can find more information about probes in the [kernel documentation](https://docs.kernel.org/trace/kprobes.html), icluding the difference between kprobes and kretprobes.
```

File: docs/book/programs/probes.md [hunk 5]
```diff
@@ -1,3 +1,52 @@
...
+To illustrate kprobes with Aya, let's write a program which
+attaches a eBPF handler to the [`tcp_connect`](https://elixir.bootlin.com/linux/latest/source/net/ipv4/tcp_output.c#L3837) function and allows printing the source and destination IP addresses from the socket parameter.
+
+## Design
+
+For this demo program, we are going to rely on aya-log to print IP addresses from the BPF program and not going to have any custom BPF maps (besides those created by aya-log).
+
+## eBPF code
+- From the `tcp_connect` signature, we see that `struct sock *sk` is the only function parameter. We will access it from the `ProbeContext` ctx handle.
+-  We call `bpf_probe_read_kernel` helper to copy the `struct sock_common	__sk_common` portion of the socket structure.  (For uprobe programs, we would need to call `bpf_probe_read_kernel` instead.)
```
Comment (vadorovsky) [inline docs/book/programs/probes.md:22]:
```suggestion
- We call `bpf_probe_read_kernel` helper to copy the `struct sock_common	__sk_common` portion of the socket structure.  (For uprobe programs, we would need to call `bpf_probe_read_user` instead.)
```

**Review by vadorovsky — CHANGES_REQUESTED** [review]

**Review by vadorovsky — APPROVED** [review]

**Pull request merged**